### ISSUE TITLE
Access arguments?

### ISSUE THREAD

**Issue by isiahmeadows** [samuelgoto/proposal-block-params#11]
In many DSLs, it's convenient to have the ability to pass arguments to the callback as well - a good example of this is with arguments for a makefile task, but also, most of the binding-related stuff could be addressed similarly by simply accepting arguments.

My first instinct would be something like this (obvious Ruby inspiration):

```js
forEach(items) do (item) {
    // ...
}
```

**Comment by samuelgoto** [comment]
Yep, indeed. I left this as an extension here:

https://github.com/samuelgoto/proposal-block-params#bindings

WDYT?

**Comment by isiahmeadows** [comment]
Eh...it makes sense in some contexts (like iteration), but not in others (makefile task arguments).

I did read that *after* I filed this, but that idea really didn't feel right, especially if you have more than one parameter passed to the block (consider task name + dependencies).

**Comment by samuelgoto** [comment]
Yep, agreed that the syntax needs more work. Can you help me identify what are the most common use cases where you'd want arguments?

> especially if you have more than one parameter passed to the block (consider task name + dependencies).

fwiw, in the current formulation,multiple parameters could be set:

```javascript
foreach ({key, value} in map) {
  // ...
}
```

**Comment by isiahmeadows** [comment]
The most common use cases would be in iteration-like contexts, but not all iteration is synchronous (like above). One example where multiple distinct arguments would be nice would be with native `forEach`, which would work out-of-the-box with explicit parameters:

```js
list.forEach do (item, i) {
    // Do something that requires both the current item and its index.
}
```

> fwiw, in the current formulation,multiple parameters could be set:

To clarify, I'm speaking about the callee's side, not the callback's side:

```js
task("foo", ["dep1", "dep2"]) do (args) {
    // ...
}
```

-----

While looking at porting some of my own code (a WIP DOM framework) over to use this, I ran into one case, although your binding proposal would work:

```js
// Original
r.dynamic(stream, (r, item) => {
    // re-render on each received item
})

// This proposal
dynamic(stream) do (item) {
    // re-render on each received item
}

// Binding proposal
dynamic(item in stream) {
    // ...
}
```

Also, previously, I had my components defined like this:

```js
// Original
const Component = m.component((r, attrs, children) => {
    // ...
})

// What it could be
const Component = m.component do (attrs, children) {
    // ...
}
```

**Comment by samuelgoto** [comment]
> The most common use cases would be in iteration-like contexts, but not all iteration is synchronous 
> (like above). One example where multiple distinct arguments would be nice would be with native 
> forEach, which would work out-of-the-box with explicit parameters:

```
list.forEach do (item, i) {
    // Do something that requires both the current item and its index.
}
```

In the current formulation, this is what the syntax would look like:

```javascript
foreach ({item, index} in list) {
    // Do something that requires both the current item and its index.
}
```

Would that work?

**Comment by samuelgoto** [comment]
> WIP DOM framework

I would love to hear more about your DOM framework :) This was very much started because I feel in love with [Kotlin's examples](https://kotlinlang.org/docs/reference/type-safe-builders.html) of DSLs for DOM construction so hoping it will help this area.

Care to share any links?

**Comment by isiahmeadows** [comment]
@samuelgoto 

> In the current formulation, this is what the syntax would look like: [...]

I'm aware. I was responding with that awareness in mind.

> Care to share any links?

It's currently private and purely local, since I'm only just now getting to the beginning stages of implementing the logic, and haven't implemented any tests (after spending literal months figuring out what I wanted the API to be).

Here's an example ~~ripped off of~~based on [Elm's radio buttons example](http://elm-lang.org/examples/radio-buttons):

```js
import * as m from "my-framework/hyperscript"
import marked from "marked"

const Sizes = {
    Small: "0.8em",
    Medium: "1em",
    Large: "1.2em",
}

const Toggle = m.component((r, {change, name}) => {
    r.h("label[style.padding=20px]", {onclick() { change.send(name) }}, r => {
        r.h("input[type=radio][name=font-size]")
        r.t(name)
    })
})

const MyApp = m.component((r, {update}) => {
    const change = m.ref.cell("Medium")

    r.h("fieldset", r => {
        r.c(Toggle, {change, name: "Small"})
        r.c(Toggle, {change, name: "Medium"})
        r.c(Toggle, {change, name: "Large"})
    })

    r.h("div", {
        style: {fontSize: change.pipe(
            m.ref.map(name => Sizes[name])
        )},
        props: {innerHTML: update.pipe(
            m.ref.map(update => marked(update.content))
        )},
    })
})

m.render(document.getElementById("app"), r => {
    const Intro = `
# Anna Karenina

## Chapter 1

Happy families are all alike; every unhappy family is unhappy in its own way.

Everything was in confusion in the Oblonskys’ house. The wife had discovered
that the husband was carrying on an intrigue with a French girl, who had been a
governess in their family, and she had announced to her husband that she could
not go on living in the same house with him...
`

    r.c(MyApp, {update: m.ref.of({content: Intro})})
})
```

**Comment by isiahmeadows** [comment]
If I were to write it using this proposal, I'd probably do something like this:

```js
import * as m from "my-framework/block"
import marked from "marked"

const Sizes = {
    Small: "0.8em",
    Medium: "1em",
    Large: "1.2em",
}

const Toggle = m.component do ({change, name}) {
    h("label[style.padding=20px]", {onclick() { change.send(name) }}) {
        h("input[type=radio][name=font-size]") {}
        t(name) {}
    }
}

const MyApp = m.component do ({update}) {
    const change = m.ref.cell("Medium")

    h("fieldset") {
        c(Toggle, {change, name: "Small"}) {}
        c(Toggle, {change, name: "Medium"}) {}
        c(Toggle, {change, name: "Large"}) {}
    }

    h("div", {
        style: {fontSize: change.pipe(
            m.ref.map(name => Sizes[name])
        )},
        props: {innerHTML: update.pipe(
            m.ref.map(update => marked(update.content))
        )},
    }) {}
}

m.render(document.getElementById("app")) {
    const Intro = `
# Anna Karenina

## Chapter 1

Happy families are all alike; every unhappy family is unhappy in its own way.

Everything was in confusion in the Oblonskys’ house. The wife had discovered
that the husband was carrying on an intrigue with a French girl, who had been a
governess in their family, and she had announced to her husband that she could
not go on living in the same house with him...
`

    c(MyApp, {update: m.ref.of({content: Intro})}) {}
}
```

Kotlin bindings, you could imagine, would look similar.

**Comment by samuelgoto** [comment]
What about the following?

```javascript
const Toggle = m.component do ({change, name}) {
    label({style: {padding: 20px}}]) {
      ::onclick() { change.send(name) }})
        
       input({type: radio, name: font-size}) {
       }
       
       span(name) {}
    }
}
```

**Comment by isiahmeadows** [comment]
I've considered going a similar route to that, except I ran into a few glitches:

1. It's unclear how to integrate custom components.
2. It's unclear how to integrate custom events.
3. I'd rather avoid introducing the overhead of a proxy for `this` to discern the difference between the two.

Of course, this is more of a hypothetical front-end, and I'm developing a shared backend you can write the front-end stuff based on. It is similar to incremental-dom and virtual-dom in that regard, and you *could* write your own front-end that works that way.

Mind if you could [find me on Gitter](https://gitter.im/isiahmeadows/), and we can talk more on this privately? I'd rather avoid polluting the issue further on this topic, since it's not even public yet.

**Comment by littledan** [comment]
Arguments seem pretty important to me, especially with how you have examples with arguments (e.g., under "C#'s foreach").

Bikeshed syntax: maybe it's odd, but could we reuse the arrow function arrow?

```
foreach(list) item => {
}
```

Otherwise I like the `foreach (map) { |key, value| }` syntax.

**Comment by samuelgoto** [comment]
I dislike ruby's ```foreach (map) { |key, value| }``` syntax, but I think 

``` javascript
foreach(list) item => {
}
```

or 

``` javascript
foreach(list) do (item) {
}
```

Is growing on me.

May I ask what are the first impressions / reservations for the ```in``` syntax? Doesn't that look more consistent with the current usage of

```
for (a in list) {
}
```

Don't you think it feels more natural to re-use that syntax/construct for params?

```
foreach (a in list) {
}
```

?

**Comment by littledan** [comment]
I like either of those too.

My immediate reaction to `in` is a little hesitant. It makes me think of a for-in loop, where things would be different in two ways:
- You'd be assigning to an existing binding, rather than making a new binding (BAD--use a let)
- You'd be iterating over object keys and up the prototype chain (BAD--use for-of)

Even though it's not supposed to be any of those things, it raises my subconscious JS linter flags.

**Comment by samuelgoto** [comment]
just as another data point, @erights also suggested the following syntax:

```javascript
foreach(list) do (item) {
}
```

Which I think is a reasonable form and seems to map better to people's intuition than my original form. I'm going to update the text to reflect that.

**Comment by littledan** [comment]
SGTM
